src/grappler/ngraph_optimizer.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#ifndef NGRAPH_TF_NGRAPHOPTIMIZER_H_
+#define NGRAPH_TF_NGRAPHOPTIMIZER_H_
+
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/graph/graph.h"
+#include "tensorflow/core/graph/graph_constructor.h"
+#include "tensorflow/core/graph/graph_def_builder.h"
+#include "tensorflow/core/grappler/optimizers/custom_graph_optimizer.h"
+
+#include "ngraph_api.h"
+#include "ngraph_assign_clusters.h"
+#include "ngraph_capture_variables.h"
+#include "ngraph_deassign_clusters.h"
+#include "ngraph_encapsulate_clusters.h"
+#include "ngraph_log.h"
+#include "ngraph_mark_for_clustering.h"
+#include "ngraph_rewrite_for_tracking.h"
+#include "tf_graph_writer.h"
+
+#include <iomanip>
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+// Custom Grappler Optimizer for NGraph-TF
+class NgraphOptimizer : public tensorflow::grappler::CustomGraphOptimizer {
+ public:
+  NgraphOptimizer() = default;
+  ~NgraphOptimizer() override = default;
+
+  string name() const override { return "NgraphOptimizer"; };
+
+  Status Init(
+      const tensorflow::RewriterConfig_CustomGraphOptimizer* config) override {
+    return Status::OK();
+  }
+
+  Status Optimize(tensorflow::grappler::Cluster* cluster,
+                  const tensorflow::grappler::GrapplerItem& item,
+                  GraphDef* output) override;
+
+  void Feedback(tensorflow::grappler::Cluster* cluster,
+                const tensorflow::grappler::GrapplerItem& item,
+                const GraphDef& optimize_output, double result) override;
+
+ protected:
+  void DumpGraphs(Graph& graph, int idx, std::string filename_prefix,
+                  std::string title);
+
+ private:
+  static std::string DotFilename(std::string kind, int idx) {
+    return GraphFilenamePrefix(kind, idx) + ".dot";
+  }
+
+  static std::string PbtxtFilename(std::string kind, int idx) {
+    return GraphFilenamePrefix(kind, idx) + ".pbtxt";
+  }
+
+  static std::string GraphFilenamePrefix(std::string kind, int idx) {
+    std::stringstream ss;
+    ss << kind << "_" << std::setfill('0') << std::setw(4) << idx;
+#if defined NGRAPH_DISTRIBUTED
+    ngraph::Distributed dist;
+    int Rank_ID = dist.get_rank();
+    ss << "_" << std::setfill('0') << std::setw(4) << Rank_ID;
+#endif
+    return ss.str();
+  }
+
+  static std::string GraphFilenamePrefix(std::string kind, int idx,
+                                         int sub_idx) {
+    std::stringstream ss;
+    ss << GraphFilenamePrefix(kind, idx) << "_" << std::setfill('0')
+       << std::setw(4) << sub_idx;
+#if defined NGRAPH_DISTRIBUTED
+    ngraph::Distributed dist;
+    int Rank_ID = dist.get_rank();
+    ss << "_" << std::setfill('0') << std::setw(4) << Rank_ID;
+#endif
+    return ss.str();
+  }
+
+  static int FreshIndex() {
+    mutex_lock l(s_serial_counter_mutex);
+    return s_serial_counter++;
+  }
+
+  static bool DumpAllGraphs() {
+    return std::getenv("NGRAPH_TF_DUMP_GRAPHS") != nullptr;
+  }
+
+  static bool DumpPrecaptureGraphs() {
```

Comment:
These functions like `DumpPrecaptureGraphs` etc are also present in `class NGraphRewritePass : public GraphOptimizationPass`. They do not use any class variables etc, and might be moved out of both classes to utils, so that we avoid this duplication

Reply:
Done